File: folkmq/src/main/java/org/noear/folkmq/client/MqMessage.java
package org.noear.folkmq.client;

import org.noear.folkmq.common.MqAssert;
import org.noear.socketd.utils.StrUtils;

import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

/**
 * 消息
 *
 * @author noear
 * @since 1.0
 */
public class MqMessage implements IMqMessage {
    private String sender;
    private String tid;
    private String tag;
    private String content;
    private Date scheduled;
    private Date expiration;
    private boolean sequence;
    private int qos = 1;
    protected Map<String, String> attrMap = new HashMap<>();

    protected MqTransaction transaction;

    public MqMessage(String content) {
        this.tid = StrUtils.guid();
        this.content = content;
    }

    /**
     * 发送者
     * */
    @Override
    public String getSender() {
        return sender;
    }

    /**
     * 跟踪ID
     */
    @Override
    public String getTid() {
        return tid;
    }

    /**
     * 标签
     * */
    @Override
    public String getTag() {
        return tag;
    }

    /**
     * 内容
     */
    @Override
    public String getContent() {
        return content;
    }

    /**
     * 定时派发
     */
    public Date getScheduled() {
        return scheduled;
    }

    /**
     * 过期时间
     */
    @Override
    public Date getExpiration() {
        return expiration;
    }

    /**
     * 是否有事务
     */
    public boolean isTransaction() {
        return transaction != null;
    }

    /**
     * 是否为顺序
     */
    @Override
    public boolean isSequence() {
        return sequence;
    }

    /**
     * 质量等级（0 或 1）
     */
    public int getQos() {
        return qos;
    }

    public MqMessage tag(String tag) {
        this.tag = tag;
        return this;
    }

    /**
     * 定时派发
     */
    public MqMessage scheduled(Date scheduled) {
        this.scheduled = scheduled;
        return this;
    }

    /**
     * 过期时间
     */
    public MqMessage expiration(Date expiration) {
        this.expiration = expiration;
        return this;
    }

    /**
     * 是否事务（内部接口）
     */
    public MqMessage transaction(MqTransaction transaction) {
        if (transaction != null) {
            this.transaction = transaction;
            transaction.binding(this);
        }

        return this;
    }

    protected String tmid() {
        if (transaction == null) {
            return null;
        } else {
            return transaction.tmid();
        }
    }

    protected MqMessage internalSender(String sender) {
        this.sender = sender;
        return this;
    }

    /**
     * 是否顺序
     */
    public MqMessage sequence(boolean sequence) {
        this.sequence = sequence;
        return this;
    }

    /**
     * 质量等级（0 或 1）
     */
    public MqMessage qos(int qos) {
        this.qos = qos;
        return this;
    }

    /**
     * 属性获取
     */
    @Override
    public String getAttr(String name) {
        return attrMap.get(name);
    }

    /**
     * 获取属性字典
     */
    public Map<String, String> getAttrMap() {
        return Collections.unmodifiableMap(attrMap);
    }

    /**
     * 属性配置
     */
    public MqMessage attr(String name, String value) {
        MqAssert.assertMeta(name, "name");
        MqAssert.assertMeta(value, "value");

        attrMap.put(name, value);
        return this;
    }
}

File: folkmq-lang-js/folkmq/src/client/MqMessage.ts
import {StrUtils} from "@noear/socket.d/utils/StrUtils";
import {MqTransaction} from "./MqTransaction";

export interface MqMessageBase {
    /**
     * 发送人
     * */
    getSender(): string | null;

    /**
     * 跟踪ID
     */
    getTid(): string;

    /**
     * 标签
     * */
    getTag(): string | null;

    /**
     * 内容
     * */
    getBody(): ArrayBuffer;

    /**
     * 过期时间 //::long
     */
    getExpiration(): Date | null;

    /**
     * 是否事务
     */
    isTransaction(): boolean;

    /**
     * 是否有序
     */
    isSequence(): boolean;

    /**
     * 质量等级（0 或 1） //:int
     */
    getQos(): number;

    /**
     * 获取属性
     */
    getAttr(name: string): string | null;
}

export class MqMessage implements MqMessageBase {
    private readonly _tid: string;
    private readonly _body: ArrayBuffer;

    private _sender: string | null = null;
    private _tag: string | null = null;
    private _scheduled: Date | null = null;
    private _expiration: Date | null = null;
    private _sequence: boolean = false;
    private _sequenceSharding: string | null = null;
    private _qos: number = 1;

    protected _attrMap = new Map<string, string>;
    protected _transaction: MqTransaction;

    constructor(body: string | ArrayBuffer, tid?: string) {
        if (tid) {
            this._tid = tid;
        } else {
            this._tid = StrUtils.guid();
        }

        if (body instanceof ArrayBuffer) {
            this._body = body;
        } else {
            this._body = StrUtils.strToBuf(body);
        }
    }

    getSender(): string | null {
        return this._sender;
    }

    getTid(): string {
        return this._tid;
    }

    getTag(): string | null {
        return this._tag;
    }

    getBody(): ArrayBuffer {
        return this._body;
    }

    getScheduled(): Date | null {
        return this._scheduled;
    }

    getExpiration(): Date | null {
        return this._expiration;
    }

    isTransaction(): boolean {
        return this._transaction != null;
    }

    isSequence(): boolean {
        return this._sequence;
    }

    getSequenceSharding(): string | null {
        return this._sequenceSharding;
    }

    getQos(): number {
        return this._qos;
    }

    tag(tag: string): MqMessage {
        this._tag = tag;
        return this;
    }

    asJson(): MqMessage {
        this.attr("Content-Type", "application/json");
        return this;
    }

    scheduled(scheduled: Date): MqMessage {
        this._scheduled = scheduled;
        return this;
    }


    expiration(expiration: Date): MqMessage {
        this._expiration = expiration;
        return this;
    }

    sequence(sequence: boolean, sharding?: string): MqMessage {
        this._sequence = sequence;
        if (sharding) {
            this._sequenceSharding = (sequence ? sharding : null);
        }
        return this;
    }

    transaction(transaction: MqTransaction): MqMessage {
        if (transaction) {
            this._transaction = transaction;
            transaction.binding(this);
        }

        return this;
    }

    getTmid(): string | null {
        if (this._transaction == null) {
            return null;
        } else {
            return this._transaction.tmid();
        }
    }

    internalSender(sender: string): MqMessage {
        this._sender = sender;
        return this;
    }

    qos(qos: number): MqMessage {
        this._qos = qos;
        return this;
    }

    getAttr(name: string): string | null {
        let tmp = this._attrMap.get(name);
        return tmp ? tmp : null;
    }

    getAttrMap(): Map<string, string> {
        return this._attrMap;
    }

    attr(name: string, value: string): MqMessage {
        this._attrMap.set(name, value);
        return this;
    }
}